streamalert/rules_engine/rules_engine.py

```diff
@@ -213,15 +213,82 @@ def _rule_analysis(self, payload, rule):
 
         return alert
 
-    def _configure_outputs(self, rule):
+    def _configure_outputs(self, record, rule):
+        """Configure the outputs for the rule
+
+        Args:
+            record (dict): Record to pass through to dynamic_outputs
+            rule (rule.Rule): Attributes for the rule which triggered the alert
+        Returns:
+            set: unique set of outputs, only required outputs if the rule is staged
+        """
         # Check if the rule is staged and, if so, only use the required alert outputs
         if rule.is_staged(self._rule_table):
             all_outputs = self._required_outputs_set
         else:  # Otherwise, combine the required alert outputs with the ones for this rule
             all_outputs = self._required_outputs_set.union(rule.outputs_set)
+            if rule.dynamic_outputs:
+                dynamic_outputs = self._configure_dynamic_outputs(record, rule)
+                all_outputs = all_outputs.union(dynamic_outputs)
 
         return all_outputs
 
+    @classmethod
+    def _configure_dynamic_outputs(cls, record, rule):
+        """Generate list of outputs from dynamic_outputs
+
+        Args:
+            record (dict): Record to pass through to the dynamic_output function
+            rule (rule.Rule): Attributes for the rule which triggered the alert
+        Returns:
+            list: list of additional outputs to append to the current set
+        """
+        args_list = [record]
+        if rule.context:
+            # Pass context to dynamic_output function if context exists
+            args_list.append(rule.context)
+
+        additional_outputs = []
+        for dynamic_output in rule.dynamic_outputs_set:
+            outputs = cls._call_dynamic_output(dynamic_output, rule.name, args_list)
+            if not outputs:
+                continue
+
+            for output in outputs:
+                if ":" not in output:
```

Comment:
Minor nitpick: Recommend putting this logic into a function, maybe call it `_is_valid_output_name()` or something, to make it clear why this logic is here.

Reply:
Great point!